### PR TITLE
feat: `to_gbq` without a destination table writes to a temporary table

### PR DESCRIPTION
**Pull request by tswast**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed) In fact, should have improved some as I added unit tests for a couple related `DataFrame.to_gbq` input validation branches.
- [x] Appropriate docs were updated (if necessary) Yes, including code samples that act as system tests :-)

Fixes internal issue 307942243 🦕

### CODE REVIEW

File: third_party/bigframes_vendored/pandas/core/frame.py [hunk 1]
```diff
@@ -125,29 +125,49 @@ def to_numpy(self, dtype=None, copy=False, na_value=None, **kwargs) -> np.ndarra
...
             >>> # destination_table = PROJECT_ID + "." + DATASET_ID + "." + TABLE_NAME
             >>> df.to_gbq("bigframes-dev.birds.test-numbers", if_exists="replace")
+            'bigframes-dev.birds.test-numbers'
+
+        Write a DataFrame to a temporary BigQuery table in the anonymous dataset.
+
+            >>> df = bpd.DataFrame({'col1': [1, 2], 'col2': [3, 4]})
+            >>> destination = df.to_gbq(ordering_id="ordering_id")
+            >>> # The table created can be read outside of the current session.
+            >>> bpd.reset_session()  # For demonstration, only.
```
Comment (shobsi) [inline third_party/bigframes_vendored/pandas/core/frame.py:153]: Is my impression correct that we prefer `close_session`, and `reset_session` was added back only for backward compatibility?

File: third_party/bigframes_vendored/pandas/core/frame.py [hunk 2]
```diff
@@ -125,29 +125,49 @@ def to_numpy(self, dtype=None, copy=False, na_value=None, **kwargs) -> np.ndarra
...
-            destination_table (str):
+            destination_table (Optional[str]):
                 Name of table to be written, in the form ``dataset.tablename``
                 or ``project.dataset.tablename``.
 
-            if_exists (str, default 'fail'):
+                If no ``destination_table`` is set, a new temporary table is
+                created in the BigQuery anonymous dataset.
+
+            if_exists (Optional[str], default 'fail'):
```
Comment (shobsi) [inline third_party/bigframes_vendored/pandas/core/frame.py:170]: Should update the default to ``None``

File: tests/unit/test_pandas.py
Comment (shobsi) [inline tests/unit/test_pandas.py:1]: Is the change in this file intended? a type annotation and a const change

File: bigframes/session/__init__.py [hunk 3]
```diff
@@ -223,6 +223,17 @@ def _create_and_bind_bq_session(self):
         query_job.result()  # blocks until finished
         self._session_id = query_job.session_info.session_id
 
+        # The anonymous dataset is used by BigQuery to write query results and
+        # session tables. BigQuery DataFrames also writes temp tables directly
+        # to the dataset, no BigQuery Session required. Note: there is a
```
Comment (shobsi) [inline bigframes/session/__init__.py:228]: Could you please help elaborate "no BigQuery Session required", isn't a session always there in BigQuery DataFrames?

Comment (tswast) [reply bigframes/session/__init__.py:228]: It is, but I'm working on removing that. This is a step in that direction. :-)

File: tests/unit/test_pandas.py
Comment (tswast) [reply tests/unit/test_pandas.py:1]: Yes. This test was failing, so this change helped me debug why. The `SELECT 1` that was there before made me think the session opening logic was running, but it was really the query from this test.

File: third_party/bigframes_vendored/pandas/core/frame.py (hunk 1 above)
Comment (tswast) [reply third_party/bigframes_vendored/pandas/core/frame.py]: Done.

**Review by shobsi — APPROVED** [review]

**Pull request merged**